python/src/iceberg/conversions.py

```diff
@@ -0,0 +1,267 @@
+def _(primitive_type, partition_value_as_str: str):
+    return bytes(partition_value_as_str, "UTF-8")
+
+
+@from_partition_value_to_py.register(DecimalType)
+def _(primitive_type, partition_value_as_str: str):
+    return Decimal(partition_value_as_str)
+
+
+@singledispatch
+def to_bytes(primitive_type, value) -> bytes:
+    """A generic function which converts a built-in python value to bytes
+
+    Args:
+        primitive_type(PrimitiveType): An implementation of the PrimitiveType base class
+        value: The value to convert to bytes (The type of this value depends on which dispatched function is
+            used--check dispatchable functions for typehints)
+    """
+    raise TypeError(f"Cannot serialize value, type {primitive_type} not supported: '{value}'")
+
+
+@to_bytes.register(BooleanType)
+def _(primitive_type, value: bool) -> bytes:
+    return struct.pack("<?", 1 if value else 0)
+
+
+@to_bytes.register(IntegerType)
+@to_bytes.register(DateType)
+def _(primitive_type, value: int) -> bytes:
+    return struct.pack("<i", value)
+
+
+@to_bytes.register(LongType)
+@to_bytes.register(TimeType)
+@to_bytes.register(TimestampType)
+@to_bytes.register(TimestamptzType)
+def _(primitive_type, value: int) -> bytes:
+    return struct.pack("<q", value)
+
+
+@to_bytes.register(FloatType)
+def _(primitive_type, value: float) -> bytes:
+    """
+    Note: float in python is implemented using a double in C. Therefore this involves a conversion of a 32-bit (single precision)
+    float to a 64-bit (double precision) float which introduces some imprecision.
+    """
+    return struct.pack("<f", value)
+
+
+@to_bytes.register(DoubleType)
+def _(primitive_type, value: float) -> bytes:
+    return struct.pack("<d", value)
+
+
+@to_bytes.register(StringType)
+def _(primitive_type, value: str) -> bytes:
+    return value.encode("UTF-8")
+
+
+@to_bytes.register(UUIDType)
+def _(primitive_type, value: uuid.UUID) -> bytes:
+    return struct.pack(">QQ", (value.int >> 64) & 0xFFFFFFFFFFFFFFFF, value.int & 0xFFFFFFFFFFFFFFFF)
+
+
+@to_bytes.register(BinaryType)
+@to_bytes.register(FixedType)
+def _(primitive_type, value: bytes) -> bytes:
+    return value
+
+
+@to_bytes.register(DecimalType)
+def _(primitive_type, value: Decimal) -> bytes:
+    value_as_tuple = value.as_tuple()
+    value_precision = len(value_as_tuple.digits)
+    value_scale = -value_as_tuple.exponent
+
+    if value_scale != primitive_type.scale:
+        raise ValueError(f"Cannot serialize value, scale of value does not match type {primitive_type}: {value_scale}")
+    elif value_precision > primitive_type.precision:
+        raise ValueError(
+            f"Cannot serialize value, precision of value is greater than precision of type {primitive_type}: {value_precision}"
+        )
+
+    with localcontext() as ctx:
+        ctx.prec = primitive_type.precision
+        unscaled_value = int((value * 10**primitive_type.scale).to_integral_value(rounding=ROUND_HALF_DOWN))
```

Comment:
nit, I don't think rounding should matter here since there shouldn't be anything after the decimal point?

Reply:
Good point, removed the rounding. Thanks!